src/board.rs

```diff
@@ -0,0 +1,1021 @@
+
+    pub fn get(&self, position: Coordinate) -> Result<Square, &str> {
+        if position.y < 0 || position.x < 0 {
+            return Err("negative coordinates");
+        };
+        let x = position.x as usize;
+        let y = position.y as usize;
+
+        if y >= self.squares.len() {
+            Err("y-coordinate is too large for board height") // TODO: specify the coordinate and height
+        } else if x >= self.squares[0].len() {
+            Err("x-coordinate is too large for board width") // TODO: specify the coordinate and width
+        } else {
+            match self.squares[y][x] {
+                None => Err("Invalid position"),
+                Some(square) => Ok(square),
+            }
+        }
+    }
+
+    pub fn set(&mut self, position: Coordinate, player: usize, value: char) -> Result<(), &str> {
+        if position.y < 0 || position.x < 0 {
+            return Err("negative coordinates");
+        };
+        let x = position.x as usize;
+        let y = position.y as usize;
+
+        if player >= self.roots.len() {
+            Err("player does not exist") // TODO: specify the number of players and which player this is
+        } else if y >= self.squares.len() {
+            Err("y-coordinate is too large for board height") // TODO: specify the coordinate and height
+        } else if x >= self.squares[0].len() {
+            Err("x-coordinate is too large for board width") // TODO: specify the coordinate and width
+        } else {
+            match self.squares[y][x] {
+                Some(_) => {
+                    self.squares[y][x] = Some(Square::Occupied(player, value));
+                    Ok(())
+                }
+                None => Err("Can't set the value of a non-existant square"),
+            }
+        }
+    }
+
+    // TODO: safety on index access like get and set - ideally combine error checking for all 3
+    pub fn clear(&mut self, position: Coordinate) {
+        if self.squares[position.y as usize][position.x as usize].is_some() {
+            self.squares[position.y as usize][position.x as usize] = Some(Square::Empty);
+        }
+    }
+
+    pub fn truncate(&mut self) {
+        let mut attatched = HashSet::new();
+        for root in self.roots.iter() {
+            let tree = self.depth_first_search(*root);
+            for bit in tree {
+                attatched.insert(bit);
+            }
+        }
+
+        for y in 0..self.height() {
+            for x in 0..self.width() {
+                let c = Coordinate {
+                    x: x as isize,
+                    y: y as isize,
+                };
+                if !attatched.contains(&c) {
+                    self.clear(c);
+                }
+            }
+        }
+    }
+
+    pub fn get_root(&self, player: usize) -> Result<Coordinate, &str> {
+        if player >= self.roots.len() {
+            Err("Invalid player")
+        } else {
+            Ok(self.roots[player])
+        }
+    }
```

Comment:
Similar to other suggestions around `get`:

```suggestion
    pub fn get_root(&self, player: usize) -> Result<Coordinate, &str> {
        match self.roots.get(player) {
            Some(player) => Ok(*player),
            None => Err("Invalid player"),
        }
    }
```